### DISCUSSION TITLE
2.0 error on relationship annotations missing Relationship[...]

### DISCUSSION THREAD

**Discussion by phy1729** [sqlalchemy/sqlalchemy#8691] [Usage Questions] [answered]
The below code runs without warning under 1.4.42 even with `SQLALCHEMY_WARN_20=1`; however under 2.0.0b2 an exception is raised `sqlalchemy.exc.ArgumentError: Type annotation for "Foo.bars" should use the syntax "Mapped[list]" or "Relationship[list]".  To leave the attribute unmapped, use ClassVar[list], assign a value to the attribute, or set __allow_unmapped__ = True on the class.`

From the migration page I had assumed that running without warning with `SQLALCHEMY_WARN_20=1` meant the code would not error with 2.0. Is this a missing warning?

Another complication is `sqlalchemy.orm.Relationship` (for the corrected annotation) does not exist in 1.4.42.

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship


Base = declarative_base()


class Foo(Base):
    __tablename__ = 'foo'

    id = Column(Integer, primary_key=True)

    bars: list['Bar'] = relationship('Bar', back_populates='foo')


class Bar(Base):
    __tablename__ = 'bar'

    id = Column(Integer, primary_key=True)

    foo: Foo = relationship(Foo, back_populates='bars', cascade='all')
```

**Comment by zzzeek** [comment, marked as answer]
hi -

unfortunately 2.0 can't support the given syntax above as it now has meaning within the declarative system.    your mapping will work if you apply the ``__allow_unmapped__`` attribute to either a common base class, or to each class directly:

```py
class B:
    __allow_unmapped__ = True

Base = declarative_base(cls=B)
```


The 2.0 declarative annotation style was not developed yet when the SQLALCHEMY_WARN_20 system was implemented, so while we could possibly add new warnings to the Mypy plugin under 1.4, it might not be worth the effort, as the Mypy plugin doesn't take place when the mappings are actually run in any case.   1.4 does not do any runtime inspection of annotations and it would need to have a large infrastructure of python 3 typing oriented code to do so, which would not be something very feasible to add within this very late stage for the 1.4 series.

1.4's Mypy plugin supports the ``Mapped[]`` syntax as well, so for a 1.4/2.0 approach that supports mypy, so the best solution is to use ``Mapped[]`` explicitly with the relationship:

```py
from sqlalchemy import Column, ForeignKey
from sqlalchemy import Integer
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Mapped

Base = declarative_base()


class Foo(Base):
    __tablename__ = 'foo'

    id = Column(Integer, primary_key=True)

    bars: Mapped[list['Bar']] = relationship('Bar', back_populates='foo')


class Bar(Base):
    __tablename__ = 'bar'

    id = Column(Integer, primary_key=True)
    foo_id = Column(ForeignKey("foo.id"))

    foo: Mapped[Foo] = relationship(Foo, back_populates='bars', cascade='all')

```

**Reply by zzzeek** [reply]
the error message is working pretty badly here so I'm going to fix that up.

**Reply by zzzeek** [reply]
you may need to have access to the fixes I will be making in #8692 for this to work fully.

**Comment by zzzeek** [comment]
this will be added to the [main migration guide](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html) as an explciit step that must be taken.

**Reply by phy1729** [reply]
Thank you!
